### PR TITLE
Upgrade Vulkan Headers to 1.4

### PR DESCRIPTION
**Pull request by nickdriscoll-lunarg**
This PR upgrades the Vulkan headers to v1.4.303

### CODE REVIEW

File: framework/generated/khronos_generators/khronos_base_generator.py [hunk 1]
```diff
@@ -461,7 +461,38 @@ def beginFeature(self, interface, emit):
 
         # Some generation cases require that extra feature protection be suppressed
         if self.genOpts.protect_feature:
-            self.featureExtraProtect = self.get_feature_protect(interface)
+            self.featureExtraProtect = self.__get_feature_protect(interface)
```
Comment (jzulauf-lunarg) [inline framework/generated/khronos_generators/khronos_base_generator.py:464]: Where is this function defined?  It's not on dev, and it's not in this PR

Comment (jzulauf-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py:464]: I inserted a raise Exception on the line above it, and it's never hit.  Are we sure we understand what is happening in this code, or are we missing a commit in the PR?

**Review by jzulauf-lunarg — CHANGES_REQUESTED** [review]
There appears to be stray code or ??  A function that is never defined is also never called.

My latest codegen appears to rebase cleanly (except for needing to rerun generate_vulkan on one change).

File: framework/generated/khronos_generators/khronos_base_generator.py (hunk 1 above)
Comment (bradgrantham-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py:464]: The GitHub Ubuntu Release job runs the generators, which succeeded, and there was no difference with the output checked into the branch.  Is it possible this code isn't run any more?

Comment (bradgrantham-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py:464]: To add just a hair more context, `__get_feature_protect` wasn't executed as far back as ceecc6829, from Oct 2021.  So this is probably a function orphaned in some ancient PR that we should delete, but not as part of this PR.

**Review by mikes-lunarg — CHANGES_REQUESTED** [review]
This looks pretty close, just a handful items. The commit history isn't very interesting so I would personally prefer one squashed commit onto dev when it goes in.

File: framework/generated/khronos_generators/vulkan_generators/vulkan_enum_to_json_body_generator.py [hunk 2]
```diff
@@ -181,11 +181,12 @@ def make_decls(self):
                 if len(self.enumEnumerants[enum]):
                     body += '    switch (value) {{\n'
                     for enumerant in self.enumEnumerants[enum]:
-                        body += textwrap.indent(prefix='        ', text=textwrap.dedent('''\
-                        case {0}:
-                            jdata = "{0}";
-                            break;
-                        '''.format(enumerant)))
+                        if self.enumEnumerants[enum][enumerant]:
```
[the remark this answers is not in the document]
Comment (nickdriscoll-lunarg) [reply framework/generated/khronos_generators/vulkan_generators/vulkan_enum_to_json_body_generator.py]: This was leftover from an earlier attempt at fixing codegen. Removed.

File: framework/generated/khronos_generators/khronos_base_generator.py (hunk 1 above)
Comment (nickdriscoll-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py]: Got it. I've changed it back.

File: framework/encode/custom_vulkan_api_call_encoders.cpp [hunk 3]
```diff
@@ -294,6 +294,46 @@ VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplateKHR(VkCommandBuffer
...
+    auto pData_unwrapped      = UnwrapDescriptorUpdateTemplateInfoHandles(
+        info, pPushDescriptorSetWithTemplateInfo->pData, handle_unwrap_memory);
+
+    vulkan_wrappers::GetDeviceTable(commandBuffer)
+        ->CmdPushDescriptorSetWithTemplate2KHR(commandBuffer, pPushDescriptorSetWithTemplateInfo);
+
+    CustomEncoderPostCall<format::ApiCallId::ApiCall_vkCmdPushDescriptorSetWithTemplate2KHR>::Dispatch(
+        manager, commandBuffer, pPushDescriptorSetWithTemplateInfo);
+}
+
```
[the remark this answers is not in the document]
Comment (nickdriscoll-lunarg) [reply framework/encode/custom_vulkan_api_call_encoders.cpp]: Of course, my bad. They're changed now,

File: framework/encode/custom_vulkan_api_call_encoders.cpp [hunk 4]
```diff
@@ -294,6 +294,46 @@ VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplateKHR(VkCommandBuffer
         manager, commandBuffer, descriptorUpdateTemplate, layout, set, pData);
 }
 
+VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplate2(
```
[the remark this answers is not in the document]
Comment (nickdriscoll-lunarg) [reply framework/encode/custom_vulkan_api_call_encoders.cpp:341]: Added.

File: framework/encode/custom_vulkan_api_call_encoders.h [hunk 5]
```diff
@@ -43,6 +43,9 @@ VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplateKHR(VkCommandBuffer
                                                                uint32_t                   set,
                                                                const void*                pData);
 
+VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplate2(
+    VkCommandBuffer commandBuffer, const VkPushDescriptorSetWithTemplateInfoKHR* pPushDescriptorSetWithTemplateInfo);
```
[the remark this answers is not in the document]
Comment (nickdriscoll-lunarg) [reply framework/encode/custom_vulkan_api_call_encoders.h]: Fixed.

File: framework/generated/khronos_generators/vulkan_generators/blacklists.json [hunk 6]
```diff
@@ -30,6 +31,7 @@
     "VkDescriptorImageInfo",
     "VkWriteDescriptorSet",
     "VkPerformanceValueINTEL",
+    "VkPushDescriptorSetWithTemplateInfo",
     "VkPushDescriptorSetWithTemplateInfoKHR",
```
[the remark this answers is not in the document]
Comment (nickdriscoll-lunarg) [reply framework/generated/khronos_generators/vulkan_generators/blacklists.json]: Ok, that makes sense. Removed.

File: framework/encode/custom_vulkan_api_call_encoders.cpp [hunk 7]
```diff
@@ -294,6 +337,45 @@ VKAPI_ATTR void VKAPI_CALL CmdPushDescriptorSetWithTemplateKHR(VkCommandBuffer
...
+
+        manager->EndCommandApiCallCapture(commandBuffer);
+    }
+
+    auto handle_unwrap_memory = manager->GetHandleUnwrapMemory();
+    auto pData_unwrapped      = UnwrapDescriptorUpdateTemplateInfoHandles(
+        info, pPushDescriptorSetWithTemplateInfo->pData, handle_unwrap_memory);
+
+    vulkan_wrappers::GetDeviceTable(commandBuffer)
+        ->CmdPushDescriptorSetWithTemplate2KHR(commandBuffer, pPushDescriptorSetWithTemplateInfo);
```
Comment (mikes-lunarg) [inline framework/encode/custom_vulkan_api_call_encoders.cpp:373]: KHR

File: framework/encode/custom_vulkan_api_call_encoders.cpp [hunk 8]
```diff
@@ -250,6 +250,49 @@ VKAPI_ATTR void VKAPI_CALL UpdateDescriptorSetWithTemplate(VkDevice
...
+        EncodeDescriptorUpdateTemplateInfo(manager, encoder, info, pData);
+
+        manager->EndCommandApiCallCapture(commandBuffer);
+    }
+
+    auto handle_unwrap_memory = VulkanCaptureManager::Get()->GetHandleUnwrapMemory();
+    auto pData_unwrapped      = UnwrapDescriptorUpdateTemplateInfoHandles(info, pData, handle_unwrap_memory);
+
+    vulkan_wrappers::GetDeviceTable(commandBuffer)
+        ->CmdPushDescriptorSetWithTemplateKHR(commandBuffer, descriptorUpdateTemplate, layout, set, pData_unwrapped);
```
Comment (mikes-lunarg) [inline framework/encode/custom_vulkan_api_call_encoders.cpp:290]: KHR

File: framework/generated/khronos_generators/khronos_base_generator.py [hunk 9]
```diff
@@ -463,6 +463,37 @@ def beginFeature(self, interface, emit):
...
+            group_name = element.get('name')
+            self.enum_names.add(group_name)
+            enumerants = dict()
+            for elem in element.findall('enum'):
+                name = elem.get('name')
+                value = elem.get('value')
+                if value:
+                    enumerants[name] = value
+            self.enumEnumerants[group_name] = enumerants
+
```
Comment (mikes-lunarg) [inline framework/generated/khronos_generators/khronos_base_generator.py:496]: I'm curious why this entire block is part of this PR, it doesn't seem to affect the vulkan output...

Comment (mikes-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py:496]: @jzulauf-lunarg is this what you meant by stray code? https://github.com/LunarG/gfxreconstruct/pull/1920#pullrequestreview-2512985931

Comment (nickdriscoll-lunarg) [reply framework/generated/khronos_generators/khronos_base_generator.py:496]: I'm not 100% sure how this code got here. I imagine I messed something up during the rebase I had to perform as a part of this PR. From what I can tell from the git history, this code _should_ just be removed.

**Review by mikes-lunarg — APPROVED** [review]

**Review by bradgrantham-lunarg — APPROVED** [review]

**Pull request merged**